pkg/epp/flowcontrol/registry/registry.go

```diff
@@ -0,0 +1,848 @@
+			state.activeQueueEmptyOnShards[shardID] = true
+		}
+		newShards[i] = shard
+	}
+
+	// 2. Add the new shards to the active list.
+	fr.activeShards = append(fr.activeShards, newShards...)
+
+	// 3. Re-partition the config for all active shards (old and new).
+	return fr.repartitionShardConfigsLocked()
+}
+
+// scaleDownLocked handles marking shards for graceful draining and re-partitioning.
+// It expects the registry's write lock to be held.
+func (fr *FlowRegistry) scaleDownLocked(newTotalActive int) error {
+	currentActive := len(fr.activeShards)
+
+	fr.logger.Info("Scaling down shards", "currentActive", currentActive, "newTotalActive", newTotalActive)
+
+	// Identify the shards to drain. These are the ones at the end of the active list.
+	// The slice from index `newTotalActive` to the end contains the shards to move.
+	shardsToDrain := fr.activeShards[newTotalActive:]
+
+	// Update the active list to only include the remaining active shards.
+	fr.activeShards = fr.activeShards[:newTotalActive]
+
+	// Move them to the draining list and mark them.
+	fr.drainingShards = append(fr.drainingShards, shardsToDrain...)
+	for _, shard := range shardsToDrain {
+		shard.markAsDraining()
+	}
+
+	// Re-partition the config across the remaining active shards.
+	return fr.repartitionShardConfigsLocked()
+}
+
+// applyFlowSynchronizationLocked is the "commit" step of `RegisterOrUpdateFlow`.
+// It updates the central `flowState` and propagates the changes to all shards.
+// It expects the registry's write lock to be held.
+func (fr *FlowRegistry) applyFlowSynchronizationLocked(
+	spec types.FlowSpecification,
+	policy framework.IntraFlowDispatchPolicy,
+	queues []framework.SafeQueue,
+) {
+	flowID := spec.ID
+	state, exists := fr.flowStates[flowID]
+	var oldPriority uint
+	isPriorityChange := false
+
+	// Get the combined list of all shards for state updates and propagation.
+	allShards := fr.getAllShardsLocked()
+
+	if !exists {
+		// This is a new flow.
+		state = newFlowState(spec, allShards)
+		fr.flowStates[flowID] = state
+	} else {
+		// This is an update to an existing flow.
+		if state.spec.Priority != spec.Priority {
+			isPriorityChange = true
+			oldPriority = state.spec.Priority
+		}
+		state.update(spec, allShards)
+	}
+
+	// Propagate the update to all shards (active and draining).
+	// The `queues` slice must align exactly with the `allShards` slice.
+	if len(allShards) != len(queues) {
+		// This indicates a severe logic error during the prepare/commit phase synchronization (a race in
+		// `RegisterOrUpdateFlow`).
+		panic(fmt.Sprintf(
+			"invariant violation: shard count (%d) and prepared queue count (%d) mismatch during commit for flow %s",
+			len(allShards), len(queues), flowID))
+	}
+
+	for i, shard := range allShards {
+		shard.synchronizeFlow(spec, policy, queues[i])
+	}
+
+	// If this was a priority change, attempt to GC the newly-draining queue immediately if it's already empty.
```

Comment:
Do we need this if the default GC mechanism will clean this up anyway? this looks out of place here.

Reply:
This highly depends on the flows rate of change, it seems we are assuming it to be high, why is that?

Reply:
This entire optimization is no longer relevant with the new composite `FlowKey` design, which treats each `(ID, Priority)` pair as a distinct, independent flow. There is no longer a special "priority change" or "migration" event; the old flow simply becomes Idle as traffic shifts, and the new generational GC handles it cleanly and efficiently.
This simplification completely resolves the concern about designing for a high rate of change. The new model is much simpler.